File: python/074. Search a 2D Matrix.py
#coding=utf-8
__author__ = 'xuxuan'
class Solution(object):
    def searchMatrix(self, matrix, target):
        """
        :type matrix: List[List[int]]
        :type target: int
        :rtype: bool
        """
        row=len(matrix)
        if row==0:
            return False
        col=len(matrix[0])
        begin=0
        end=row*col
        while begin<end:
            middle=(begin+end)/2
            x=middle/col
            y=middle%col
            if matrix[x][y]==target:
                return True
            else:
                if matrix[x][y]>target:
                    end=middle
                else:
                    begin=middle+1
        return False

File: java/074. Search a 2D Matrix.java
public class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int begin=0;
        int end=matrix.length-1;
        int t=end/2;
        if (target<matrix[0][0]||target>matrix[matrix.length-1][matrix[matrix.length-1].length-1]) {
            return false;
        }
        while (end-begin>1) {
            if (matrix[t][0]==target) {
                return true;
            }else if (matrix[end][0]<target) {
                t=end;
                begin=end;
                break;
            }else if (matrix[t][0]>target) {
                end=t;
                t=(begin+end)/2;
            }else {
                begin=t;
                t=(begin+end)/2;
            }
        }
        if (matrix[t][0]==target||matrix[begin][0]==target||matrix[end][0]==target) {
            return true;
        }else if (matrix[t][0]>target) {
            t=t-1;
        }
        int b=0;
        int e=matrix[t].length-1;
        int tt=(b+e)/2;
        while (e-b>1) {
            if (matrix[t][tt]==target) {
                return true;
            }else if (matrix[t][tt]>target) {
                e=tt;
                tt=(b+e)/2;
            }else {
                b=tt;
                tt=(b+e)/2;
            }
        }
        if (matrix[t][b]==target||matrix[t][e]==target) {
            return true;
        }
        return false;
    }
}